[GSoC] physics/continuum_mechanics: Added the `draw` method for the Truss Class

AdvaitPote opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #24040 

#### Brief description of what is fixed or changed
As a final step for the `Truss` class, the `draw` method for the same has been added. This method, similar to the one in `Beam`, returns a plot object representing the state of the truss denoting all of its properties like loads and supports along with its structural components like nodes and members. 

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics
  * Added the `draw` method for getting the diagram of a given `truss`.
<!-- END RELEASE NOTES -->

AdvaitPote committed:
draw method added in Truss

AdvaitPote commented:
@Psycho-Pirate

Psycho-Pirate commented on sympy/physics/continuum_mechanics/truss.py:
```diff
@@ -733,3 +737,226 @@ def solve(self):
+            pinned support are in the +X and +Y direction while those of a
+            roller support are in the +Y direction. For the load, the range
+            of angles one can input goes all the way to 360 degrees which, in the
+            plot, is the angle that the load vector makes with positive x-axis.
```
Just add the direction of the angle as well after "positive x-axis"

Psycho-Pirate commented:
Its working fine. I was thinking if we could make it a bit more appealing. Maybe add some thickness to the members. We can discuss this.

AdvaitPote committed:
rectangles added for members along with few documentation changes

AdvaitPote committed:
minor documentation changes added

AdvaitPote commented:
I have added the changes recommended. The members have been replaced by rectangles rotated by an angle determined by the coordinates of the two nodes of the member. Also, I was thinking the doctests for the method can be a little more developed so I have replaced them with the example you had mentioned above. This truss I feel because of its symmetricity and because it is a bit more complicated, makes documentation a little better.

```
>>> from sympy.physics.continuum_mechanics.truss import Truss
>>> t = Truss()
>>> t.add_node("A", -4, 0)
>>> t.add_node("B", 0, 0)
>>> t.add_node("C", 4, 0)
>>> t.add_node("D", 8, 0)
>>> t.add_node("E", 6, 2/math.sqrt(3))
>>> t.add_node("F", 2, 2*math.sqrt(3))
>>> t.add_node("G", -2, 2/math.sqrt(3))
>>> t.add_member("AB","A","B")
>>> t.add_member("BC","B","C")
>>> t.add_member("CD","C","D")
>>> t.add_member("AG","A","G")
>>> t.add_member("GB","G","B")
>>> t.add_member("GF","G","F")
>>> t.add_member("BF","B","F")
>>> t.add_member("FC","F","C")
>>> t.add_member("CE","C","E")
>>> t.add_member("FE","F","E")
>>> t.add_member("DE","D","E")
>>> t.apply_support("A","pinned")
>>> t.apply_support("D","roller")
>>> t.apply_load("G", 3, 90)
>>> t.apply_load("E", 3, 90)
>>> t.apply_load("F", 2
>>> p = t.draw()
>>> p.show()
```
<img width="499" alt="Screenshot 2022-09-05 at 10 55 34 AM" src="https://user-images.githubusercontent.com/92469698/188367454-65c7a48f-e9d1-4777-b537-882ee1a64448.png">)

Psycho-Pirate commented:
> I have added the changes recommended. The members have been replaced by rectangles rotated by an angle determined by the coordinates of the two nodes of the member. Also, I was thinking the doctests for the method can be a little more developed so I have replaced them with the example you had mentioned above. This truss I feel because of its symmetricity and because it is a bit more complicated, makes documentation a little better.

This is looking really good. The docs are looking great as well. I was thinking that maybe we can replace the initial example truss image with the one created by the draw function as well.

Also one small nitpick, some nodes seem a bit off. I tried zooming in on nodes B and C. 
For node B it shows 
![nb](https://user-images.githubusercontent.com/50932406/188380920-1ae450f7-bf6c-452a-b574-76127367e889.png)

For node C
![nc](https://user-images.githubusercontent.com/50932406/188380949-52d1ec56-8fab-468e-8757-3f354b8bea12.png)

It looks like node B is a bit off.
Similarly, nodes D, E and F seem a bit off too. Just look into it.

Davide-sd commented:
@AdvaitPote very nice work on the `Truss` class.

In the last few days I have been trying it on different problems, specifically the ones dealing with the principal of virtual works and energy methods in order to compute the displacements of the nodes.

Sometime it is useful to provide coordinates with physical units. The symbolic part of the code dealing with internal forces computation is going to work well. However, physical units on coordinates will make the `draw` method fails, because coordinates might contain `Symbol` or `Quantity` objects, which are not numerical data. Nonetheless, visualizing a frame is extremely important to verify the user inputs.

So, I'm going to post an example of a calculation involving the `Truss` class:

```python
from sympy import *
from sympy.physics.units import m, mm, N
from sympy.physics.continuum_mechanics import Truss

def mme(*args):
    """Return a element-by-element multiplication of SymPy matrices.
    """
    if len(args) < 1:
        raise ValueError("Please, provide at least two matrices.")
    r = args[0]
    for i in range(1, len(args)):
        r = matrix_multiply_elementwise(r, args[i])
    return r

def get_lengths(t):
    """Compute the length of the members of a Truss"""
    members_lengths = dict()
    for member in list(t._members):
        start, end = t._members[member]
        members_lengths[member] = sqrt((t._node_coordinates[start][0]-t._node_coordinates[end][0])**2 + (t._node_coordinates[start][1]-t._node_coordinates[end][1])**2)
    return members_lengths

# length
L = 4 * m

def build_structure():
    t = Truss()
    t.add_node("A", 0, 0)
    t.add_node("B", L, 0)
    t.add_node("C", 2 * L, 0)
    t.add_node("D", 3 * L, 0)
    t.add_node("E", L, L)
    t.add_node("F", 2 * L, L)
    t.add_member("AB", "A", "B")
    t.add_member("BC", "B", "C")
    t.add_member("CD", "C", "D")
    t.add_member("AE", "A", "E")
    t.add_member("BE", "B", "E")
    t.add_member("BF", "B", "F")
    t.add_member("CF", "C", "F")
    t.add_member("DF", "D", "F")
    t.add_member("EF", "E", "F")
    t.apply_support("A", type="pinned")
    t.apply_support("D", type="roller")
    return t

# solve for actual forces
t1 = build_structure()
t1.apply_load("E", magnitude=40e03 * N, direction=270)
t1.apply_load("C", magnitude=100e03 * N, direction=270)
t1.solve()

# NOTE: this fails because coordinates contains a `Quantity`
# the same happens if they contain symbols...
# t1.draw().show()

# unit load
t2 = build_structure()
t2.apply_load("B", magnitude=1, direction=270)
t2.solve()

# cross sectional area
A = 1800 * mm**2
E = 200e03 * N / mm**2

L = Matrix(list(get_lengths(t1).values()))
# actual forces
F = Matrix(list(t1.internal_forces.values()))
# virtual forces
Fv = Matrix(list(t2.internal_forces.values()))

# principle of virtual work
displ = sum(mme(F, Fv, L) / (A * E)).n()
displ
# out: 0.00351524771279376m
```

As noted in the code, drawing this structure will raise an error. I propose the following:
* modify the signature of `draw(self)` to `draw(self, subs_dict=None)`, where `subs_dict` is a user-provided substitution dictionary. In the above example:  `t1.draw(subs_dict={m: 1})`. Here, the user is assigning a value of 1 to a meter, allowing the method to draw the structure.
* inside the `draw` method, extract every free symbols and `Quantity` from the nodes coordinates. Implement a (set) difference between these symbolic objects and the keys of `subs_dict`. If the length of the resulting set is greater than zero, raise a `ValueError` and inform the user that in order to proceed, a substitution dictionary is required.

Psycho-Pirate commented:
Thanks for pointing this out! @Davide-sd

AdvaitPote committed:
added member_lengths attribute

AdvaitPote commented:
I have added the `member_lengths` attribute as was requested in the issue #24040 pointed out by @Davide-sd. Also, I will be adding additional examples and tests including only Symbolics in the future commits. 

I am looking into the issues pointed out above. I have my mid-semester examinations going on so it might take some time.

AdvaitPote committed:
Merge branch 'master' into draw_truss

AdvaitPote committed:
merge conflict resolved

AdvaitPote committed:
subs_dict added to the draw method

AdvaitPote committed:
Merge branch 'draw_truss' of https://github.com/AdvaitPote/sympy into draw_truss

AdvaitPote commented:
I have added `subs_dict` to the method.

AdvaitPote committed:
rectified non-alignment of nodes

AdvaitPote commented:
@Psycho-Pirate I have made the required changes. The node misalignment was arising from the way the starting point of each member rectangle `xy` is calibrated. As per various possible cases of the initial node and the final node, modifications were required to be made to the `xy`, `width`, and `height` attributes of the member. The nodes look better now.

This is node B now. 
<img width="621" alt="Screenshot 2022-10-12 at 11 43 02 PM" src="https://user-images.githubusercontent.com/92469698/195433749-0580da51-8c81-4a68-ab96-e63189ecdacb.png">

And this is node C. 
<img width="625" alt="Screenshot 2022-10-13 at 1 04 49 AM" src="https://user-images.githubusercontent.com/92469698/195433864-1f987867-5086-4fe0-b84b-cd43d19de04a.png">

AdvaitPote commented:
@Psycho-Pirate

oscarbenjamin commented:
Closing and reopening to rerun the tests.

oscarbenjamin commented:
@Psycho-Pirate does this look good to you?

Psycho-Pirate commented:
Sorry for such a delayed response! This looks good to me! 
Please restart the tests @oscarbenjamin

AdvaitPote committed:
changes for the code quality test to pass

AdvaitPote committed:
modified arrowprops dict in _draw_loads

Psycho-Pirate commented:
The mailmap file that is present in your branch doesn't contain some names. These names are present in the master branch. I think you should just replace your mailmap with the master's mailmap and it should pass.
@AdvaitPote

AdvaitPote committed:
replaced .mailmap with master's .mailmap

Psycho-Pirate commented:
It looks like your master isn't upto date with Sympy's master

AdvaitPote committed:
Merge remote-tracking branch 'upstream/master' into draw_truss

AdvaitPote committed:
trying to pass authors test

AdvaitPote commented:
Resolved! @Psycho-Pirate
